File: artifacts/tulipindicators/decompose/src/buffer.c
ti_buffer *ti_buffer_new(int size)
{
  const int s = ((int) (sizeof(ti_buffer))) + ((size - 1) * ((int) (sizeof(double))));
  ti_buffer *ret = (ti_buffer *) malloc((unsigned int) s);
  ret->size = size;
  ret->pushes = 0;
  ret->index = 0;
  ret->sum = 0;
  return ret;
}


/*** DEPENDENCIES:
typedef struct 
{
  int size;
  int pushes;
  int index;
  double sum;
  double vals[1];
} ti_buffer
----------------------------
***/


void ti_buffer_free(ti_buffer *buffer)
{
  free(buffer);
}


/*** DEPENDENCIES:
typedef struct 
{
  int size;
  int pushes;
  int index;
  double sum;
  double vals[1];
} ti_buffer
----------------------------
***/

File: artifacts/tulipindicators/automatic_debug/src/buffer.rs
use crate::*;

pub fn ti_buffer_new(size: i32) -> Option<Box<TiBuffer>> {
	println!("\n* ti_buffer_new");
    if size <= 0 {
        return None;
    }

    Some(Box::new(TiBuffer {
        size,
        pushes: 0,
        index: 0,
        sum: 0.0,
        vals: vec![0.0; size as usize],
    }))
}
pub fn ti_buffer_free(buffer: TiBuffer) {
	println!("\n* ti_buffer_free");
    // In Rust, memory is automatically managed by the ownership system.
    // When `buffer` goes out of scope here, it will be dropped (freed).
    // No explicit free is needed as the Vec<f64> inside will be automatically deallocated.
}
